src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -329,7 +329,7 @@ address ArchiveBuilder::reserve_buffer() {
   ReservedSpace rs(buffer_size, MetaspaceShared::core_region_alignment(), os::vm_page_size());
   if (!rs.is_reserved()) {
     log_error(cds)("Failed to reserve " SIZE_FORMAT " bytes of output buffer.", buffer_size);
-    MetaspaceShared::unrecoverable_writing_error();
+    MetaspaceShared::writing_error();
```

Comment:
I don't understand how that could work. Would the subsequent access to rs.base not crash the VM?

Reply:
I agree with Thomas. We need to stop any further operations and exit the safepoint. Something like this

```
void VM_PopulateDumpSharedSpace::doit() {
  ...
  StaticArchiveBuilder builder;
  builder.gather_source_objs();
>>>>
  if (builder.reserve_buffer() == nullptr) {
     // report error ...
     this->_failed = true; // 
     return;
   }
<<<<
```

The failure needs to be propagated to the main thread.

```
  VM_PopulateDumpSharedSpace op;
  VMThread::execute(&op);
>>>>
   if (op._failed) {
       THROW_MSG(.....);
    }
<<<<
}
```

And the VM will eventually exit with an unhandled exception.